<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследование-данных-и-проверка-гипотез" data-toc-modified-id="Исследование-данных-и-проверка-гипотез-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование данных и проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Гипотеза-1:-есть-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок." data-toc-modified-id="Гипотеза-1:-есть-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Гипотеза 1: есть зависимость между количеством детей и возвратом кредита в срок.</a></span></li><li><span><a href="#Гипотеза-2:-есть-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок." data-toc-modified-id="Гипотеза-2:-есть-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Гипотеза 2: есть зависимость между семейным положением и возвратом кредита в срок.</a></span></li><li><span><a href="#Гипотеза-3:-есть-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок." data-toc-modified-id="Гипотеза-3:-есть-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Гипотеза 3: есть зависимость между уровнем дохода и возвратом кредита в срок.</a></span></li><li><span><a href="#Гипотеза-4:-разные-цели-кредита-влияют-на-его-возврат-в-срок." data-toc-modified-id="Гипотеза-4:-разные-цели-кредита-влияют-на-его-возврат-в-срок.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Гипотеза 4: разные цели кредита влияют на его возврат в срок.</a></span></li><li><span><a href="#Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Дальнейшие-шаги-для-более-детального-исследования:" data-toc-modified-id="Дальнейшие-шаги-для-более-детального-исследования:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Дальнейшие шаги для более детального исследования:</a></span></li><li><span><a href="#Вопросы-ревьюеру:" data-toc-modified-id="Вопросы-ревьюеру:-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вопросы ревьюеру:</a></span></li></ul></div>

# Исследование надежности заемщиков

## Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение, количество детей клиента, его ежемесячный доход и цель получения кредита на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

In [1]:
# импортируем библиотеку pandas
import pandas as pd

In [2]:
# читаем csv-файл

try:
    data = pd.read_csv('C:\\Users\\lyubo\\Downloads\\Практикум\\Спринт_2\\data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [3]:
# посмотрим на данные
display(data.head(20))

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
# выведем основную информацию о данных
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


## Предобработка данных

### Удаление пропусков

In [5]:
# выявим пропущенные значения для каждого столбца и посчитаем их
print(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


Сначала поработаем со столбцом total_income.

In [6]:
# перебираем каждый тип занятости в наборе уникальных значений столбца income_type
for income in data['income_type'].unique():
    # на каждом шаге цикла с помощью атрибута loc выбираем строки,
    # в которых в income_type текущий тип занятости (income) и есть пропуски в total_income
    # и записываем в них медианное значение total_income среди строк с текущим типом занятости (income)
    data.loc[(data['income_type'] == income) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == income), 'total_income'].median()


# проверка, что все пропуски в income_type заполнены
print(data['total_income'].isna().sum())

0


### Обработка аномальных значений

В данных могут встречаться артефакты, а у нас это явно отрицательное количество дней трудового стажа в столбце `days_employed`.

In [7]:
# сохраняем в исходный столбец days_employed данные по модулю
data['days_employed'] = data['days_employed'].abs()

# проверка
print(data['days_employed'])

0          8437.673028
1          4024.803754
2          5623.422610
3          4124.747207
4        340266.072047
             ...      
21520      4529.316663
21521    343937.404131
21522      2113.346888
21523      3112.481705
21524      1984.507589
Name: days_employed, Length: 21525, dtype: float64


In [8]:
# выведем медианное значение трудового стажа (в днях) для каждого типа занятости

print(data.groupby('income_type')['days_employed'].median())

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64


У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть.

In [9]:
# узнаем сколько детей у заёмщиков (уникальные значения)
print(data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


In [10]:
# удалим аномальные значения в столбце children из датафрейма
data = data.loc[(data['children'] != -1) & (data['children'] != 20)]

### Удаление пропусков (продолжение)

Теперь поработаем со столбцом days_employed.

In [11]:
# перебираем каждый тип занятости в наборе уникальных значений столба income_type
for income in data['income_type'].unique():
    # на каждом шаге цикла с помощью атрибута loc выбираем строки,
    # в которых в income_type текущий тип занятости (income) и есть пропуски в days_employed 
    data.loc[(data['income_type'] == income) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == income), 'days_employed'].median()
    # и записываем в них среднее значение total_income среди строк с текущим типом занятости (income)

# проверка, что все пропуски в days_employed заполнены
print(data['days_employed'].isna().sum())

0


In [12]:
# убедимся, что вообще все пропуски заполнены
print(data.isna().sum()) 

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


### Изменение типов данных

In [13]:
# заменим вещественный тип данных в столбце total_income на целочисленный
data['total_income'] = data['total_income'].astype('int')

### Обработка дубликатов

In [14]:
# приведём значения столбца education к нижнему регистру, чтобы избавиться от неявных дубликатов
data['education'] = data['education'].str.lower()

In [15]:
# посчитаем строки-дубликаты в датасете
print(data.duplicated().sum())

71


In [16]:
# удалим их
data = data.drop_duplicates()
print(data.duplicated().sum())

0


### Категоризация данных

<a id='income_category'></a>
Создадим столбец с категориями по уровню дохода:

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

[Перейти к выводу в п.3.3](#conclusion)

In [17]:
# для этого создадим функцию categorize_income()

def categorize_income(total_income):
    if total_income <= 30000:
        return 'E'
    elif total_income <= 50000:
        return 'D'
    elif total_income <= 200000:
        return 'C'
    elif total_income <= 1000000:
        return 'B'
    return 'A'

In [18]:
# применим функцию
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [19]:
# посмотрим на цели получения кредита, выведем уникальные значения
print(data['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


Создадим столбец с фиксированными категориями по целям получения кредита, в который войдут следующие категории:

- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [20]:
# для этого создадим функцию categorize_purpose()

def categorize_purpose(row):
    purpose = row['purpose']
    try:
        if 'авто' in purpose:
            return 'операции с автомобилем'
        if 'жиль' in purpose or 'недвиж' in purpose:
            return 'операции с недвижимостью'
        if 'свадьб' in purpose:
            return 'проведение свадьбы'
        if 'образов' in purpose:
            return 'получение образования'
    except:
        return 'другая категория'

In [21]:
# применим функцию
data['purpose_category'] = data.apply(categorize_purpose, axis=1)

# проверим себя
print(data['purpose_category'].head(15))

0     операции с недвижимостью
1       операции с автомобилем
2     операции с недвижимостью
3        получение образования
4           проведение свадьбы
5     операции с недвижимостью
6     операции с недвижимостью
7        получение образования
8           проведение свадьбы
9     операции с недвижимостью
10    операции с недвижимостью
11    операции с недвижимостью
12          проведение свадьбы
13      операции с автомобилем
14    операции с недвижимостью
Name: purpose_category, dtype: object


Посмотрим на датасет после преобразований.

In [22]:
display(data.head(20))

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0  365286.622650         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              среднее             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        118514   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   
15   женат / замужем                 0      F   сотрудник     0        116820   
16   женат / замужем                 0      F   сотрудник     0        289202   
17  гражданский брак                 1      M   сотрудник     0         90410   
18    вдовец / вдова                 2      F   пенсионер     0         56823   
19         в разводе                 3      F   сотрудник     0        242831   

                                   purpose total_income_category  \
0                            покупка жилья                     B   
1                  приобретение автомобиля                     C   
2                            покупка жилья                     C   
3               дополнительное образование                     B   
4                          сыграть свадьбу                     C   
5                            покупка жилья                     B   
6                        операции с жильем                     B   
7                              образование                     C   
8                    на проведение свадьбы                     C   
9   

## Исследование данных и проверка гипотез

### Гипотеза 1: есть зависимость между количеством детей и возвратом кредита в срок.

Сначала сделаем **подробные поэтапные расчёты** для наглядности. Далее приведём два варианта кратких версий, одну из которых и будем использовать в последующих расчётах. 

In [23]:
# группируем данные по количеству детей, выбираем столбец debt и считаем
# количество всего полученных кредитов у клиентов с разным количеством детей
dep_repayment_children_number = data.groupby('children')['debt'].count()

# для каждой группы по количечеству детей считаем количество задолженностей по кредиту
dep_repayment_children_sum = data.groupby('children')['debt'].sum()

# для каждой группы по количеству детей считаем среднее арифметическое по задолженности по кредиту
dep_repayment_children_mean = dep_repayment_children_sum / dep_repayment_children_number

print(dep_repayment_children_number)
print()
print(dep_repayment_children_sum)
print()
# округлим результат до тысячных
print(dep_repayment_children_mean.round(3))

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

children
0    0.075
1    0.092
2    0.095
3    0.082
4    0.098
5    0.000
Name: debt, dtype: float64


**Краткий вариант 1:** те же подсчёты мы могли сделать более оптимальным путём, введя к группировке агрегирующую функцию. Её и будем использовать в последующих подсчётах.

In [24]:
dep_repayment_children = data.groupby('children').agg({'debt': ['count', 'sum', 'mean']})

print(dep_repayment_children.round(3))

           debt             
          count   sum   mean
children                    
0         14091  1063  0.075
1          4808   444  0.092
2          2052   194  0.095
3           330    27  0.082
4            41     4  0.098
5             9     0  0.000


**Краткий вариант 2:** те же подсчёты мы могли сделать более оптимальным путём, создав сводную:

In [25]:
data_pivot_repayment_children = data.pivot_table(index = ['children'], values = 'debt', aggfunc = ['count', 'sum', 'mean'])
print(data_pivot_repayment_children.round(3))

          count   sum   mean
           debt  debt   debt
children                    
0         14091  1063  0.075
1          4808   444  0.092
2          2052   194  0.095
3           330    27  0.082
4            41     4  0.098
5             9     0  0.000


**Выводы:**

<u>Основные:</u> 
- если особое внимание обратить именно на самые большие категории по количеству детей, а именно: 0, 1 и 2 ребёнка, мы увидим, что по мере увеличения у клиентов количества детей, увеличивается и вероятность  задолженности по кредиту. 
- клиенты с 5-ю детьми вовсе не имеют задолжностей, но т.к. выборка по клиентам с 5-ю детьми крайне мала, она не показательна.

<u>Побочный:</u>  по мере увеличения у заёмщиков количества детей, их склонность брать кредит падает. 

**Итого:** можно сделать вывод, что зависимость между количеством детей и возвратом кредита в срок есть.

### Гипотеза 2: есть зависимость между семейным положением и возвратом кредита в срок.

In [26]:
# для каждой группы по семейному положению посчитаем количество полученных кредитов (count), 
# сколько было кредитных задолженностей (sum) и среднее арифметическое (mean)

dep_repayment_status = data.groupby('family_status').agg({'debt': ['count', 'sum', 'mean']})

print(dep_repayment_status.round(3))

                        debt            
                       count  sum   mean
family_status                           
Не женат / не замужем   2796  273  0.098
в разводе               1189   84  0.071
вдовец / вдова           951   63  0.066
гражданский брак        4134  385  0.093
женат / замужем        12261  927  0.076


**Вывод:** 
- опираясь на полученные данные мы видим, что состоящие или состоявшие в браке заемщики реже допускают просрочки.

**Итого:** можно сделать вывод, что есть зависимость между семейным положением и возвратом кредита в срок.

### Гипотеза 3: есть зависимость между уровнем дохода и возвратом кредита в срок.

In [27]:
# для каждой группы по уровню дохода посчитаем количество полученных кредитов (count), 
# сколько было кредитных задолженностей (sum) и среднее арифметическое (mean)

dep_repayment_income = data.groupby('total_income_category').agg({'debt': ['count', 'sum', 'mean']})

print(dep_repayment_income.round(3))

                        debt             
                       count   sum   mean
total_income_category                    
A                         25     2  0.080
B                       5014   354  0.071
C                      15921  1353  0.085
D                        349    21  0.060
E                         22     2  0.091


[Памятка: категории ежемесячного дохода](#income_category)

**Вывод:** <a id='conclusion'></a>

- если обратиться к двум наиболее многочисленным выборкам, а именно B и С, мы видим, что по мере уменьшения уровня дохода клиента, вероятность вернуть кредит в срок значительно уменьшается.

**Итого:** можно сделать вывод, что есть ярко выраженная зависимость между уровнем дохода и возвратом кредита в срок.

### Гипотеза 4: разные цели кредита влияют на его возврат в срок.

In [28]:
# для каждой группы по цели получения кредита посчитаем количество полученных кредитов (count), 
# сколько было кредитных задолженностей (sum) и среднее арифметическое (mean)
dep_repayment_purpose = data.groupby('purpose_category').agg({'debt': ['count', 'sum', 'mean']})

print(dep_repayment_purpose.round(3))

                           debt            
                          count  sum   mean
purpose_category                           
операции с автомобилем     4279  400  0.093
операции с недвижимостью  10751  780  0.073
получение образования      3988  369  0.093
проведение свадьбы         2313  183  0.079


**Вывод:** 

<u>Основной:</u>
- мы видим, что клиенты, цель которых произвести операции с автомобилем или получить образование, чаще остальных имеют задолжность по кредиту. У тех клиентов, кто нацелен производить операции с недвижимостью или провести свадьбу, этот показатель чуть лучше.

<u>Побочный:</u>
- самой частой целью получения кредита является операция с недвижимостью, а самой редкой - проведение свадьбы. 

**Итого:** можно сделать вывод, что есть ярко выраженная зависимость между целью получения кредита и возвратом кредита в срок.

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Пропуски были выявлены в столбце `общий трудовой стаж в днях` и `ежемесячный доход`, причём показатели идентичны. Это наводит на предположение, что клиент не трудоустроен официально и, как следствие, данные по трудовому стажу и доходу не поступают.
Также такого рода пропуски, предположительно, могут говорить о том, что клиент - женщина в декрете, которая ещё не успела получить опыт работы. 

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* заполнить пропуски медианным значением — лучшее решение для количественных переменных, чтобы не исказить результат средними показателями, т.к. данные неоднородны и некоторые значения дохода и/ или трудового стажа могут сильно выделяться среди большинства.

## Общий вывод

Для исследования нам была передана статистика о платёжеспособности 21525 клиентов банка.
Статистика содержала информацию о количество детей в семье клиента, его общем трудовом стаже, возрасте, уровне образования, семейном положении, поле клиента, типе занятости, имел ли он задолженность по возврату кредитов, ежемесячном доходе, а также данные о цели получения кредита.

На стадии предобработки в данных были обнаружены и обработны пропуски, артефакты и дубликаты. Из дополнительных преобразований, для упрощения работы мы ввели категоризацию по уровню дохода клиентов и сделали более структурной категоризацию по цели получения кредита на основе общей смысловой нагрузки. 

В ходе исследования данных все выдвинутые нами гипотезы были подтверждены:

- по мере увеличения у клиентов количества детей, увеличивается и вероятность задолженности по кредиту;
- состоящие или состоявшие в браке заемщики реже допускают просрочки;
- по мере уменьшения уровня дохода клиента, вероятность вернуть кредит в срок значительно уменьшается;
- клиенты, цель которых произвести операции с автомобилем или получить образование, чаще остальных имеют задолжность по кредиту. У тех клиентов, кто нацелен производить операции с недвижимостью или провести свадьбу, этот показатель чуть лучше.

## Дальнейшие шаги для более детального исследования:

1. Исследовать зависимость между количеством детей, самейным статусом и наличием задолжностей по возврату кредита.
2. Исследовать зависимость между уровнем ежемесячных доходов, количеством детей и наличием задолжностей по возврату кредита.
3. Исследовать зависимость между количеством детей, целью получения кредита и наличием задолжностей по возврату кредита.